optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,320 @@
+    initial_params: np.ndarray,
+    initial_fval: float,
+    continuous_indices: np.ndarray,
+    lengthscale: np.ndarray,
+    tol: float,
+) -> tuple[np.ndarray, float, bool]:
+    """
+    This function optimizes the acquisition function using preconditioning.
+    Preconditioning equalizes the variances caused by each parameter and
+    speeds up the convergence.
+
+    In Optuna, acquisition functions use Matern 5/2 kernel, which is a function of `x / l`
+    where `x` is `normalized_params` and `l` is the corresponding lengthscales.
+    Then acquisition functions are a function of `x / l`, i.e. `f(x / l)`.
+    As `l` has different values for each param, it makes the function ill-conditioned.
+    By transforming `x / l` to `zl / l = z`, the function becomes `f(z)` and has
+    equal variances w.r.t. `z`.
+    So optimization w.r.t. `z` instead of `x` is the preconditioning here and
+    speeds up the convergence.
+    As the domain of `x` is [0, 1], that of `z` becomes [0, 1/l].
+    """
+    if len(continuous_indices) == 0:
+        return (initial_params, initial_fval, False)
+    normalized_params = initial_params.copy()
+
+    def negative_acqf_with_grad(scaled_x: np.ndarray) -> tuple[float, np.ndarray]:
+        # Scale back to the original domain, i.e. [0, 1], from [0, 1/s].
+        normalized_params[continuous_indices] = scaled_x * lengthscale
+        (fval, grad) = eval_acqf_with_grad(acqf_params, normalized_params)
+        # Flip sign because scipy minimizes functions.
+        # Let the scaled acqf be g(x) and the acqf be f(sx), then dg/dx = df/dx * s.
+        return (-fval, -grad[continuous_indices] * lengthscale)
+
+    scaled_cont_x_opt, neg_fval_opt, info = so.fmin_l_bfgs_b(
+        func=negative_acqf_with_grad,
+        x0=normalized_params[continuous_indices] / lengthscale,
+        bounds=[(0, 1 / s) for s in lengthscale],
+        pgtol=math.sqrt(tol),
+        maxiter=200,
+    )
+
+    if -neg_fval_opt < initial_fval or info["nit"] == 0:
+        # `nit` is the number of iterations.
+        return (initial_params, initial_fval, False)
+    else:
+        normalized_params[continuous_indices] = scaled_cont_x_opt * lengthscale
+        return (normalized_params, -neg_fval_opt, True)
+
+
+def _exhaustive_search(
+    acqf_params: AcquisitionFunctionParams,
+    initial_params: np.ndarray,
+    initial_fval: float,
+    param_idx: int,
+    choices: np.ndarray,
+) -> tuple[np.ndarray, float, bool]:
+    choices_except_current = choices[choices != initial_params[param_idx]]
+
+    all_params = np.repeat(initial_params[None, :], len(choices_except_current), axis=0)
+    all_params[:, param_idx] = choices_except_current
+    fvals = eval_acqf_no_grad(acqf_params, all_params)
+    best_idx = np.argmax(fvals)
+
+    if fvals[best_idx] > initial_fval:
+        return (all_params[best_idx, :], fvals[best_idx], True)
+    else:
+        return (initial_params, initial_fval, False)
```

Comment:
I just followed the gradient ascent where we early-return no improvement.
```suggestion
    if fvals[best_idx] > initial_fval:  # Improved.
        return (all_params[best_idx, :], fvals[best_idx], True)

    return (initial_params, initial_fval, False)  # No improvement.
```